### ISSUE TITLE
Feature Request: expose parsing apis

### ISSUE THREAD

**Issue by DanielFGray** [sql-formatter-org/sql-formatter#753]
it would be amazing if the parsing apis were exported, so other projects can take advantage of them
personally i would love to see the parser being broken into a separate package, but i can understand the maintenance overhead that might require

willing to head up a PR if this is acceptable

**Comment by nene** [comment]
Well, do you personally have a project that would like to take advantage of it, or is it more of an abstract request?

The things is, the parser isn't really much of a parser (as in, producing a proper AST), it's very specifically tailored for the needs of the formatter.

If you're looking for an SQL parser, you might instead want to consider [sql-parser-cst](https://github.com/nene/sql-parser-cst), which I specifically created to overcome the massive limitations that the "parser" in sql-formatter has. It's used by [prettier-plugin-sql-cst](https://github.com/nene/prettier-plugin-sql-cst) to format SQL in much better way than will ever be possible with the architecture of SQL Formatter.

**Comment by DanielFGray** [comment]
> do you personally have a project that would like to take advantage of it

[I do](https://github.com/danielfgray/postgres-playground). I tried sql-parser-cst but it doesn't seem to work for anything more than very simple sql commands (see [here](https://github.com/graphile/starter/blob/v5/%40app/db/migrations/committed/000001.sql) for an example of something more advanced)

I can open parsing issues in that repo if you're interested

**Comment by nene** [comment]
Well, it does support all the syntax of SQLite and BigQuery, but that's indeed very simple SQL compared to all of PostgreSQL, which will need a substantial amount of additional work before it's fully finished. It is though the most complete PostgreSQL parser implementation in JavaScript that I'm aware of. I did quite a lot of experimenting with existing ones before I finally decided to write my own.

There's an [issue that tracks the progress of Postgres implementation.](https://github.com/nene/sql-parser-cst/issues/40)

If a part that's marked as supported in there is not working, then it's indeed worth of opening a ticket. But no real point of opening tickets about the parts of syntax for which nothing has been implemented yet. Then again I haven't even mapped out all the work that remains, like the GRANT/REVOKE syntax that's used at the beginning of that big SQL file you referenced. So it might be worthwhile to open a ticket about that, or just adding a vote for existing tickets for unimplemented stuff. That way I can possibly prioritize implementing the parts that people care the most about.
